### ISSUE TITLE
GPU Memory Leak During Layer-by-Layer Inference in QuantizedModelForCausalLM (or quantize & freeze) (Layers Not Released After Deletion)

### ISSUE THREAD

**Issue by quaternior** [huggingface/optimum-quanto#411]
Hello, I have been using the library successfully.
Although I understand that it is no longer actively developed, I am currently encountering an issue where GPU memory usage explodes when performing layer-by-layer inference of an LLM. Even after explicitly deleting each layer, the memory usage continues to increase, which suggests that references may not be properly unlinked during the deletion process.
Since this problem does not occur with a non-quantized model, I believe it is unlikely to be an issue with my code logic itself.
The relevant code snippet is as follows:

```
org_model = AutoModelForCausalLM.from_pretrained(model_name, device_map="cuda")
model = QuantizedModelForCausalLM.quantize(
    org_model,
    weights=weight_dtype,
    exclude="lm_head"
)

for i in range(len(model.model.layers)):
    layer = model.model.layers[i]
    hidden_states = layer(
        hidden_states=hidden_states,
        attention_mask=attention_mask,
        position_ids=position_ids,
        past_key_values=past_key_value,
        output_attentions=output_attentions,
        use_cache=use_cache,
        cache_position=cache_position,
        position_embeddings=position_embeddings,
    )
    model.model.layers[i] = None

    gc.collect()
    torch.cuda.empty_cache()
```

**Comment by quaternior** [comment]
Found the reason. It is because of autograd which is true in default, so it is resolved with eval mode
